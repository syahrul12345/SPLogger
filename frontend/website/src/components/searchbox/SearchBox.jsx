import React, { useState } from 'react';
import { Grid, TextField, Button,FormControl,InputLabel,Select,MenuItem } from '@material-ui/core'

// Redux
import { connect } from 'react-redux';
import { UpdateSearchedAppliances } from '../../redux-modules/user/actions'
// API
import { SubmitSearch } from '../../utils/utils';

// Componenets
import ErrorDialog from '../errordialog';

const SearchBox = (props) => {
  const [searchTerm,setSearchTerm] = useState('');
  const [selectedFilter,setSelectedFilter] = useState('SerialNumber');
  const [openErrorDialog,setOpenErrorDialog] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  const searchTermHandler = () => event => {
    setSearchTerm(event.target.value)
  }
  
  const searchCallback = (resp) => {
    const { data } = resp
    if(data.length === 0) {
      setErrorMessage("No objects were found")
      setOpenErrorDialog(true)
    }else {
      // Update search results into redux
      props.dispatch(UpdateSearchedAppliances(data))
    }
  }
  const searchFailCallback = (errResp) => {
    setErrorMessage("No objects were found")
    setOpenErrorDialog(true)
  }
  const search = () => {
    SubmitSearch(searchCallback,searchFailCallback,selectedFilter,searchTerm)
  }
  const handleFilterChange = (event) => {
    setSelectedFilter(event.target.value)
  }
  const clearSearch = () => {
    props.dispatch(UpdateSearchedAppliances([]))
  }
  return (
    <Grid container spacing={2}>
      <Grid item xs={12}>
        <TextField 
          id="standard-basic" 
          label="Search for appliance" 
          InputLabelProps={{
            shrink: true,
            style: { color: '#fff' },
          }}
          style={{width:'100%'}}
          type={ selectedFilter === 'DateBought' ? 'date' : 'text'}
          onChange={searchTermHandler()}/>
          
      </Grid>
      <Grid item xs={12}>
      <FormControl>
          <InputLabel id="demo-simple-select-label">Filter</InputLabel>
          <Select
            labelId="demo-simple-select-label"
            id="demo-simple-select"
            value={selectedFilter}
            onChange={handleFilterChange}
          >
            <MenuItem value={'SerialNumber'}> SerialNumber</MenuItem>
            <MenuItem value={'Brand'}>Brand</MenuItem>
            <MenuItem value={'Model'}>Model</MenuItem>
            <MenuItem value={'Status'}>Status</MenuItem>
            <MenuItem value={'DateBought'}>Date Purchased</MenuItem>
          </Select>
        </FormControl>
      </Grid>
      
      <Grid item xs={12}>
        <Button variant="contained" color="primary" onClick={() => search()}> Search </Button>
        <Button variant="contained" color="primary" onClick={() => clearSearch()} style={{marginLeft:'5px'}}> Clear Search </Button>
      </Grid>
      <ErrorDialog open={openErrorDialog} setOpen={setOpenErrorDialog} errorMessage={errorMessage}/>
    </Grid>
  )
}
const mapStateToProps = (state) => {
  return ({
    searchResults: state.user.searchResults,
  })
}

const ConnectedSearchBox = connect(mapStateToProps)(SearchBox);

export { SearchBox, ConnectedSearchBox }