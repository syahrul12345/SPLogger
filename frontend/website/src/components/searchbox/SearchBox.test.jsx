import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { TextField, FormControl,MenuItem} from '@material-ui/core';
import { SearchBox } from './SearchBox';

describe("Render Search tests",() => {
  test("Compare snapshot",() => {
    const component = renderer.create(
      <SearchBox/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
  test("Test for the presence of 1 search box compoennt ", () => {
    const component = shallow(<SearchBox/>)
    expect(component.find(TextField).length).toEqual(1)
  })
  test("Test for the presence of 1 filter selection dropdown component",() => {
    const component = shallow(<SearchBox/>)
    expect(component.find(FormControl).length).toEqual(1)
  })
  test("Test for presence of serial number filter",() => {
    const component = shallow(<SearchBox/>)
    expect(component.find(MenuItem).at(0).props().value).toEqual('SerialNumber')
  })
  test("Test for presence of Brand filter",() => {
    const component = shallow(<SearchBox/>)
    expect(component.find(MenuItem).at(1).props().value).toEqual('Brand')
  })
  test("Test for presence of Model filter",() => {
    const component = shallow(<SearchBox/>)
    expect(component.find(MenuItem).at(2).props().value).toEqual('Model')
  })
  test("Test for presence of Status filter",() => {
    const component = shallow(<SearchBox/>)
    expect(component.find(MenuItem).at(3).props().value).toEqual('Status')
  })
  test("Test for presence of DateBought filter",() => {
    const component = shallow(<SearchBox/>)
    expect(component.find(MenuItem).at(4).props().value).toEqual('DateBought')
  })
  
})