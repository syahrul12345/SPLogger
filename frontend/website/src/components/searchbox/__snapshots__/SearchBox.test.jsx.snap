// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Search tests Compare snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
        data-shrink={true}
        htmlFor="standard-basic"
        id="standard-basic-label"
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        Search for appliance
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="standard-basic"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink={true}
        id="demo-simple-select-label"
      >
        Filter
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="demo-simple-select-label demo-simple-select"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="demo-simple-select"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
           SerialNumber
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="SerialNumber"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         Search 
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginLeft": "5px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         Clear Search 
      </span>
    </button>
  </div>
  <div />
</div>
`;
